util/Logger.h

```diff
@@ -1,56 +1,245 @@
 #ifndef _Logger_h_
 #define _Logger_h_
 
-#include <boost/log/trivial.hpp>
+#include <boost/log/sources/severity_channel_logger.hpp>
 #include <boost/log/expressions/keyword.hpp>
 #include <boost/log/utility/manipulators/add_value.hpp>
+#include <boost/log/sources/global_logger_storage.hpp>
+#include <boost/signals2/signal.hpp>
+
+#ifdef FREEORION_WIN32
+// Note: The is a workaround for Visual C++ non-conformant pre-processor
+// handling of empty macro arguments.
+// https://msdn.microsoft.com/en-us/library/hh567368.aspx
+// https://blogs.msdn.microsoft.com/vcblog/2017/03/07/c-standards-conformance-from-microsoft/
+#include <boost/preprocessor/control/if.hpp>
+#include <boost/preprocessor/facilities/is_empty_variadic.hpp>
+#endif
+
+#include <string>
+#include <set>
+#include <tuple>
 
 #include "Export.h"
 
 
-/** Initializes the logging system. Log to the given file.
- * If the file already exists it will be deleted. */
-FO_COMMON_API void InitLogger(const std::string& logFile, const std::string& pattern);
-
-/** Accessors for the App's logger */
-FO_COMMON_API void SetLoggerPriority(int priority);
-
+/** \file
+    \brief The logging system consists of named loggers with levels.
+
+    The system is a thin wrapper around boost::log.
+
+    The logging system is composed of a sink which writes to the log files and
+    sources, called loggers, which collect the log information while the
+    application is running.
+
+    Both the sinks and the sources use debug levels (trace, debug, info, warn and
+    error) to filter which log records are generated at the sources and which
+    log records are written at the sinks. Logs that are filtered out at either
+    the sink or the source are not generated by the source.
+
+    The intended uses of the levels are:
+    error - used for "major" unrecoverable errors which will affect game play.  Error level issues
+            need to be fixed.  Error level will probably not be turned off unless they are flooding
+            the logs.
+            Examples are: the game is about to crash, a string is missing from the stringtable, etc.
+    warn  - used for "minor", recoverable errors that will not affect game play but do indicate a
+            problem.
+            For example a missing id that can be ignored, an extra item in a container.
+    info  - used to report normal game state and progress.  This should be the default level of
```

Comment:
example might be helpful here... perhaps to contrast with debug, "For example reporting that the network disconnected" (ie. without the detail in debug / trace) ?